### PR TITLE
Adding windows to Travis CI

### LINKED ISSUE #221: Add windows to travis tests.
Given that we use a few bits and pieces such as signal handling that works differently on different platforms it’d be really useful to get Windows into our test matrix if possible.

https://mobile.twitter.com/travisci/status/1050475566770589697

### PR DESCRIPTION
**Pull request by almarklein**
Closes #221 

Resources:
* The [Travis blog post introducing Windows support](https://blog.travis-ci.com/2018-10-11-windows-early-release?utm_source=twitter&utm_medium=web&utm_campaign=windows_early_release)
* A [golang pr](https://github.com/docker-library/golang/blob/1e38d66d798f94d693509741db2c5595540b1434/.travis.yml)

Todo:

* [x] Get CI working for Windows
* [x] Fix tests on Windows
* ~~Also test on OS X - nope: no Python on OS X with Travis~~
* [x] While we're at it maybe also test on Pypy3?

### CODE REVIEW

**Comment by lovelydinosaur** [comment]
screen shot 2018-10-16 at 15 53 06

☝️🤣

**Comment by lovelydinosaur** [comment]
Thanks for taking a crack at this!

**Comment by davidism** [comment]
Might be best to stick with AppVeyor until Travis adds Python language support to their Windows environment. The config is straightforward compared to managing the install manually: https://github.com/pallets/click/blob/master/.appveyor.yml.

Alternatively, use `choco install python3` since Chocolatey is available in the environment.

**Comment by almarklein** [comment]
> ☝️🤣

Yeah ... you might wanna squash-merge this :)

> Alternatively, use choco install python3 since Chocolatey is available in the environment.

Yeah, I realized this later.

**Comment by almarklein** [comment]
progress!

**Comment by almarklein** [comment]
Ready for review!

**Comment by lovelydinosaur** [comment]
Amazing! Taking a look through now.

File: .travis.yml [hunk 1]
```diff
@@ -1,14 +1,38 @@
...
+
+matrix:
+  include:
+    # The Windows job is important (obviously) to test on a non-Unix platform.
+    # It also runs the test with a quite minimal setup (no uvloop or httptools),
+    # but it does have websockets (needed to run the websockets tests).
+    - os: windows
+      language: bash
+    # The pypy3 job runs a fully pure Python setup, which triggers a part of the
+    # auto-detect tests that the other tests don't. We cannot test websockets though.
```
Comment (lovelydinosaur) [inline .travis.yml:13]: How come we can't test websockets on the pypy3 environment? (h11 and wsproto?)

Comment (almarklein) [reply .travis.yml:13]: The websocket tests make use of `websockets` for the "client", which is not pure Python, so I've assumed its not possible to install it on Pypy.

**Comment by lovelydinosaur** [comment]
I oughta come back to having another look at this - it'd be nice to get it in.

File: tests/test_ssl.py [hunk 2]
```diff
@@ -105,6 +105,9 @@ def remove_files():
 
 def test_run(create_certfile_and_keyfile):
 
+    if sys.platform.startswith("win"):
+        pytest.skip("Skipping SSL test on Windows for now :(")
```
Comment (almarklein) [inline tests/test_ssl.py:109]: Because they hang. Have not investigated further, but I don't want it to hold back this PR.

Comment (lovelydinosaur) [reply tests/test_ssl.py:109]: That's fair enough, yup!

**Comment by lovelydinosaur** [comment]
Thanks for taking a crack at getting this back up to scratch @almarklein - soon as we're back to passing tests we should get it merged in. Let me know if you need my time on anything here.

**Comment by almarklein** [comment]
Thanks @tomchristie. There seem to be issues with the latest version of the websocket library on Windows. All tests passed when I ran it locally after I upgraded `websockets` they fail (just as CI). Looking into it now.

**Comment by almarklein** [comment]
Oh yeah, I do have a question: I saw you have a series of autoformatters lines up, including `black`. But when I run these it changes a lot of other code. I assume you've not "applied" auto-formatting to prevent breaking PR's?

**Comment by lovelydinosaur** [comment]
> I assume you've not "applied" auto-formatting to prevent breaking PR's?

We strictly enforce that the linting as been applied in the tests.
You'll need to run `./scripts/lint` if you want to apply the linting - it won't run automatically.

**Comment by lovelydinosaur** [comment]
> when I run these it changes a lot of other code.

Not sure what you mean there - all looks clean against the latest master to me.

**Comment by almarklein** [comment]
> You'll need to run `./scripts/lint`

Not on Windows ;)  But I ran the lines inside, and it changed a lot of code that I never touched. Mmm, will have a look again later.

**Comment by almarklein** [comment]
@tomchristie Green lights!

About the linting: it's the `isort` call that changes about 10 files when I run it. It also looks like there is no test for `isort` in `scripts/test`.

**Comment by lovelydinosaur** [comment]
Nice work!

**Pull request merged**